Build panics on rootless builder if no output or push flags given

**What steps did you take and what happened**
Rootless builders work, but the CLI panics due to failed assumptions trying to default to exporting to the local runtime.  In the case of a rootless builder, this isn't possible.  Only pushing, or outputting to the client are possible as mechanisms to utilize the build results.

```
% kubectl buildkit create --rootless
...
% kubectl build -t test:foo .
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4ae6817]

goroutine 1 [running]:
github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/build.Build.func3(0xc000594f00, 0xc00042c150, 0xc00048ef60, 0xc00048ef30, 0xc00012a030, 0xc00045bd18, 0xc000010008, 0xc000408f30, 0x0, 0xc00042c1e0, ...)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/build/build.go:710 +0x4e7
github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/build.Build(0x59598e0, 0xc00003c9c0, 0xc0001c00a0, 0x1, 0x1, 0xc000595520, 0x5959de0, 0xc0001145f0, 0x0, 0x0, ...)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/build/build.go:777 +0xf26
github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd.buildTargets(0x59598e0, 0xc00003c9c0, 0x5959de0, 0xc0001145f0, 0x590a1c0, 0xc00019a000, 0x590a1e0, 0xc00019a008, 0x590a1e0, 0xc00019a010, ...)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd/build.go:253 +0x2bc
github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd.runBuild(0x590a1c0, 0xc00019a000, 0x590a1e0, 0xc00019a008, 0x590a1e0, 0xc00019a010, 0x590a1c0, 0xc00019a000, 0x590a1e0, 0xc00019a008, ...)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd/build.go:229 +0x956
github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd.buildCmd.func1(0xc0003eb080, 0xc000278e10, 0x1, 0x3, 0x0, 0x0)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/pkg/cmd/build.go:302 +0x1fc
github.com/spf13/cobra.(*Command).execute(0xc0003eb080, 0xc0001a8010, 0x3, 0x3, 0xc0003eb080, 0xc0001a8010)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/vendor/github.com/spf13/cobra/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003eb080, 0xc00019a000, 0x590a1e0, 0xc00019a008)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/vendor/github.com/spf13/cobra/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/vendor/github.com/spf13/cobra/command.go:887
main.main()
	/Users/hiltgend/reno/go/src/github.com/vmware-tanzu/buildkit-cli-for-kubectl/cmd/kubectl-build/main.go:22 +0x101
```

If you explicitly specify an `--output` this overrides the default behavior, and local works properly and stores the build contents locally (push works as well)

```
kubectl build -t test:foo --output=type=local,dest=./tmp .
[+] Building 4.8s (7/7) FINISHED
 => [internal] load build definition from Dockerfile                                                                                      0.0s
 => => transferring dockerfile: 31B                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                         0.0s
 => => transferring context: 2B                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/busybox:latest                                                                         1.6s
 => [internal] load build context                                                                                                         0.0s
 => => transferring context: 24B                                                                                                          0.0s
 => [1/2] FROM docker.io/library/busybox:latest@sha256:a9286defaba7b3a519d585ba0e37d0b2cbee74ebfe590960b0b1d6a5e97d1e1d                   0.0s
 => => resolve docker.io/library/busybox:latest@sha256:a9286defaba7b3a519d585ba0e37d0b2cbee74ebfe590960b0b1d6a5e97d1e1d                   0.0s
 => CACHED [2/2] COPY cmd /bin/cmd                                                                                                        0.0s
 => exporting to client                                                                                                                   3.1s
 => => copying files 1.25MB                                                                                                               3.1s
```
**What did you expect to happen**

A default build should not panic. 

Most likely it should give a warning that the output will be stored in the builder only, and not accessible, and encourage the user to specify `--push` or `--output` with another explicit output strategy if they want to do something with their built image


**Environment Details:**

- kubectl buildkit version (use `kubectl buildkit version`)

v0.1.0

- Kubernetes version (use `kubectl version`)
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-27T00:38:11Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.14", GitCommit:"d2a081c8e14e21e28fe5bdfa38a817ef9c0bb8e3", GitTreeState:"clean", BuildDate:"2020-08-13T12:24:51Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- Where are you running kubernetes (e.g., bare metal, vSphere Tanzu, Cloud Provider xKS, etc.)

Fusion VM running Ubuntu

- Container Runtime and version (e.g. containerd `sudo ctr version` or dockerd `docker version` on one of your kubernetes worker nodes)

dockerd

**Builder Logs**
```
kubectl logs -l app=buildkit
time="2020-11-07T23:06:51Z" level=warning msg="NoProcessSandbox is enabled. Note that NoProcessSandbox allows build containers to kill (and potentially ptrace) an arbitrary process in the BuildKit host namespace. NoProcessSandbox should be enabled only when the BuildKit is running in a container as an unprivileged user."
time="2020-11-07T23:06:51Z" level=info msg="found worker \"foaqdjuhq02wsyndsgfkdqwwk\", labels=map[org.mobyproject.buildkit.worker.executor:oci org.mobyproject.buildkit.worker.hostname:buildkit-7d9cc68576-lb54v org.mobyproject.buildkit.worker.snapshotter:overlayfs], platforms=[linux/amd64 linux/386]"
time="2020-11-07T23:06:51Z" level=warning msg="rootless mode is not supported for containerd workers. disabling containerd worker."
time="2020-11-07T23:06:51Z" level=info msg="found 1 workers, default=\"foaqdjuhq02wsyndsgfkdqwwk\""
time="2020-11-07T23:06:51Z" level=warning msg="currently, only the default worker can be used."
time="2020-11-07T23:06:51Z" level=info msg="running server on /run/user/1000/buildkit/buildkitd.sock"
```
**Dockerfile**
```
FROM busybox:latest
COPY cmd /bin/cmd
ENTRYPOINT ["/bin/cmd"]
```

**Vote on this request**

This is an invitation to the community to vote on issues.  Use the "smiley face" up to the right of this comment to vote.

- :+1: "I would like to see this bug fixed as soon as possible"
- :-1: "There are more important bugs to focus on right now"

--- a/Makefile
+++ b/Makefile
@@ -3,6 +3,7 @@
 
 VERSION?=$(shell git describe --tags --first-parent --abbrev=7 --long --dirty --always)
 TEST_KUBECONFIG?=$(HOME)/.kube/config
+TEST_TIMEOUT=300s
 
 # Verify Go in PATH
 ifeq (, $(shell which go))
@@ -36,7 +37,7 @@ help:
 
 .PHONY: clean
 clean:
-	-rm -rf $(BIN_DIR) cover*.out cover.html
+	-rm -rf $(BIN_DIR) cover*.out cover*.html
 
 .PHONY: build
 build: $(BIN_DIR)/$(NATIVE_ARCH)/kubectl-buildkit $(BIN_DIR)/$(NATIVE_ARCH)/kubectl-build
@@ -70,7 +71,7 @@ test:
 integration:
 	@echo "Running integration tests with $(TEST_KUBECONFIG)"
 	@kubectl config get-contexts
-	TEST_KUBECONFIG=$(TEST_KUBECONFIG) go test $(GO_FLAGS) $(EXTRA_GO_TEST_FLAGS)  \
+	TEST_KUBECONFIG=$(TEST_KUBECONFIG) go test -timeout $(TEST_TIMEOUT) $(GO_FLAGS) $(EXTRA_GO_TEST_FLAGS)  \
 		$(GO_COVER_FLAGS) -coverprofile=./cover-int.out \
 		./integration/...
 

--- a/integration/common/basesuites.go
+++ b/integration/common/basesuites.go
@@ -3,6 +3,10 @@
 package common
 
 import (
+	"fmt"
+	"path"
+	"strings"
+
 	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
@@ -53,5 +57,43 @@ func (s *BaseSuite) TestSimpleBuild() {
 		dir,
 	)
 	err = RunBuild(args)
+	if isRootlessCreate(s.CreateFlags) {
+		require.Error(s.T(), err)
+		require.Contains(s.T(), err.Error(), "please specify")
+	} else {
+		require.NoError(s.T(), err, "build failed")
+	}
+}
+
+func isRootlessCreate(flags []string) bool {
+	for _, flag := range flags {
+		if strings.Contains(flag, "rootless") {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *BaseSuite) TestLocalOutputTarBuild() {
+	logrus.Infof("%s: Local Output Tar Build", s.Name)
+
+	dir, cleanup, err := NewSimpleBuildContext()
+	defer cleanup()
+	require.NoError(s.T(), err, "Failed to set up temporary build context")
+	args := []string{}
+	if s.Name != "buildkit" { // TODO wire up the default name variable
+		args = append(
+			args,
+			"--builder", s.Name,
+		)
+	}
+	args = append(
+		args,
+		"--tag", s.Name+"replaceme:latest",
+		fmt.Sprintf("--output=type=tar,dest=%s", path.Join(dir, "out.tar")),
+		dir,
+	)
+	err = RunBuild(args)
 	require.NoError(s.T(), err, "build failed")
+	// TODO - consider inspecting the out.tar for validity...
 }

--- a/integration/suites/rootless_test.go
+++ b/integration/suites/rootless_test.go
@@ -4,22 +4,24 @@ package suites
 
 import (
 	"testing"
-	//"github.com/stretchr/testify/suite"
-	//"github.com/vmware-tanzu/buildkit-cli-for-kubectl/integration/common"
+
+	"github.com/stretchr/testify/suite"
+	"github.com/vmware-tanzu/buildkit-cli-for-kubectl/integration/common"
 )
 
-//type rootlessSuite struct{ common.BaseSuite }
+type rootlessSuite struct{ common.BaseSuite }
 
 func TestRootlessSuite(t *testing.T) {
-	t.Skip("Skipping rootless due to bug! - should disable local storage mode automatically")
-	//common.Skipper(t)
+	common.Skipper(t)
 	//t.Parallel() // TODO - tests fail if run in parallel, may be actual race bug
-	/*
-		suite.Run(t, &rootlessSuite{
-			BaseSuite: common.BaseSuite{
-				Name:        "rootless",
-				CreateFlags: []string{"--rootless", "true"},
-			},
-		})
-	*/
+	suite.Run(t, &rootlessSuite{
+		BaseSuite: common.BaseSuite{
+			Name:        "rootless",
+			CreateFlags: []string{"--rootless", "true"},
+		},
+	})
 }
+
+// func (s *rootlessSuite) TestSimpleBuild() {
+// 	s.T().Skip("Rootless doesn't support loading to the runtime")
+// }

--- a/pkg/build/build.go
+++ b/pkg/build/build.go
@@ -364,8 +364,8 @@ func toSolveOpt(ctx context.Context, d driver.Driver, multiDriver bool, opt Opti
 			} else if driverFeatures[driver.DockerExporter] {
 				opt.Exports[i].Type = "docker"
 			} else {
-				// TODO should we allow building without load or push?
-				return nil, nil, errors.Wrap(err, "unable to determine runtime, please specify --output=type=[docker || containerd] or --push")
+				// TODO should we allow building without load or push, perhaps a new "nil" or equivalent output type?
+				return nil, nil, errors.Errorf("loading image into cluster runtime not supported by this builder, please specify --push or a client local output: --output=type=local,dest=. --output=type=tar,dest=out.tar ")
 			}
 		}
 	}

--- a/pkg/cmd/create.go
+++ b/pkg/cmd/create.go
@@ -128,7 +128,7 @@ Driver Specific Usage:
 	flags.StringVar(&options.configFile, "config", "", "BuildKit config file")
 	flags.StringArrayVar(&options.platform, "platform", []string{}, "Fixed platforms for current node")
 	flags.StringVar(&options.progress, "progress", "auto", "Set type of progress output [auto, plain, tty]. Use plain to show container output")
-	flags.StringVar(&options.image, "image", bkimage.DefaultImage, "Specify an alternate buildkit image")
+	flags.StringVar(&options.image, "image", "", fmt.Sprintf("Specify an alternate buildkit image (default: %s)", bkimage.DefaultImage))
 	flags.StringVar(&options.runtime, "runtime", "auto", "Container runtime used by cluster [auto, docker, containerd]")
 	flags.StringVar(&options.containerdSock, "containerd-sock", kubernetes.DefaultContainerdSockPath, "Path to the containerd.sock on the host")
 	flags.StringVar(&options.containerdNamespace, "containerd-namespace", kubernetes.DefaultContainerdNamespace, "Containerd namespace to build images in")

--- a/pkg/driver/kubernetes/factory.go
+++ b/pkg/driver/kubernetes/factory.go
@@ -139,7 +139,9 @@ func (d *Driver) initDriverFromConfig() error {
 			if err != nil {
 				return err
 			}
-			deploymentOpt.Image = bkimage.DefaultRootlessImage
+			if deploymentOpt.Rootless {
+				deploymentOpt.Image = bkimage.DefaultRootlessImage
+			}
 		case "loadbalance":
 			switch v {
 			case LoadbalanceSticky:
@@ -232,7 +234,6 @@ func (d *Driver) initDriverFromConfig() error {
 		//        user tries to set properties that should be in the config file
 		d.configMap = manifest.NewConfigMap(deploymentOpt, data)
 	}
-
 	return nil
 }
 

